tests/operators/test_grid_sampling_op.py

```diff
@@ -0,0 +1,268 @@
+
+
+@pytest.mark.parametrize('align_corners', ['no_align', 'align'])
+def test_grid_sampling_op_align_mode_backward_is_adjoint(align_corners):
+    """Test for different align modes ."""
+    _test_grid_sampling_op_x_backward(align_corners=align_corners == 'align')
+
+
+def _test_grid_sampling_op_x_backward(dim=3, interpolation_mode='bilinear', padding_mode='zeros', align_corners=False):
+    rng = RandomGenerator(0).float32_tensor
+    batch = (2, 3)
+    channel = (5, 7)
+    zyx_v = (7, 10, 20)[-dim:]
+    zyx_u = (9, 22, 30)[-dim:]
+    grid = rng((*batch, *zyx_v, dim), -1, 1.0)
+    input_shape = SpatialDimension(z=99 if dim == 2 else zyx_u[-3], y=zyx_u[-2], x=zyx_u[-1])
+    operator = GridSamplingOp(
+        grid,
+        input_shape=input_shape,
+        interpolation_mode=interpolation_mode,
+        padding_mode=padding_mode,
+        align_corners=align_corners,
+    )
+    u = rng((*batch, *channel, *zyx_u)).requires_grad_(True)
+    v = rng((*batch, *channel, *zyx_v)).requires_grad_(True)
+    (forward_u,) = operator(u)
+    forward_u.backward(v.detach())
+    (adjoint_v,) = operator.adjoint(v)
+    adjoint_v.backward(u.detach())
+    torch.testing.assert_close(u.grad, adjoint_v)
+    torch.testing.assert_close(v.grad, forward_u)
+
+
+def test_grid_sampling_op_gradcheck_x_forward():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    u = rng((1, 1, 3, 5)).requires_grad_(True)
+    gradcheck(lambda grid, u: GridSamplingOp(grid, input_shape=SpatialDimension(1, 3, 5))(u), (grid, u), fast_mode=True)
+
+
+def test_grid_sampling_op_gradcheck_grid_forward():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    u = rng((1, 1, 3, 5))
+    gradcheck(lambda grid, u: GridSamplingOp(grid, input_shape=SpatialDimension(1, 3, 5))(u), (grid, u), fast_mode=True)
+
+
+def test_grid_sampling_op_gradcheck_x_adjoint():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    v = rng((2, 1, 1, 2))
+    gradcheck(
+        lambda grid, v: GridSamplingOp(grid, input_shape=SpatialDimension(1, 2, 3)).adjoint(v),
+        (grid, v),
+        fast_mode=True,
+    )
+
+
+def test_grid_sampling_op_gradcheck_grid_adjoint():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    v = rng((2, 1, 1, 2))
+    gradcheck(
+        lambda grid, v: GridSamplingOp(grid, input_shape=SpatialDimension(1, 2, 3)).adjoint(v),
+        (grid, v),
+        fast_mode=True,
+    )
+
+
+def test_grid_sampling_op_errormsg_gridlastdim():
+    "Test if error message on wrong last dim is raised."
+    grid = torch.ones(1, 2, 3, 4)
+    with pytest.raises(ValueError, match='last dimension'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+
+def test_grid_sampling_op_errormsg_gridndims():
```

Comment:
better to use parametrize

Reply:
No. I could split it into two tests with each three lines. Using parameterize makes the code more complicated, as 2/3 lines would just be behind an if switch..